onnxruntime/wasm/api.h

```diff
@@ -222,4 +230,132 @@ int EMSCRIPTEN_KEEPALIVE OrtRun(ort_session_handle_t session,
+ * @param run_options handle of the run options
+ * @return int ORT error code. If not zero, call OrtGetLastError() to get detailed error message.
+ */
+int EMSCRIPTEN_KEEPALIVE OrtTrainingRunTrainStep(ort_training_session_handle_t training_session_handle,
+                                                 const ort_tensor_handle_t inputs, const size_t inputs_count,
+                                                 ort_tensor_handle_t outputs,
+                                                 const size_t outputs_count,
+                                                 OrtRunOptions* run_options = nullptr);
+
+/**
+ * Performs weight updates for the trainable parameters in the given training session using the optimizer model.
+ * @param training_handle handle of the training session
+ * @param run_options optional parameter of run options for this training step
+ * @returns ORT error code. If not zero, call OrtGetLastError() to get detailed error message.
+ */
+int EMSCRIPTEN_KEEPALIVE OrtTrainingOptimizerStep(ort_training_session_handle_t training_handle,
+                                                  const ort_run_options_handle_t run_options = nullptr);
+
+ort_tensor_handle_t EMSCRIPTEN_KEEPALIVE OrtTrainingEvalStep(ort_training_session_handle_t training_handle,
+                                                             const ort_run_options_handle_t options,
+                                                             size_t inputs_len,
+                                                             const ort_tensor_handle_t* inputs,
+                                                             size_t outputs_len,
+                                                             ort_tensor_handle_t* outputs);
+
+/**
+ * Retrieves the size of all parameters for the training state.
+ * When the trainable_only argument is true, the size is calculated for trainable params only.
+ *
+ * @param training_handle handle of the training session
+ * @param trainable_only skips non-trainable parameters when true.
+ */
+size_t* EMSCRIPTEN_KEEPALIVE OrtTrainingGetParametersSize(ort_training_session_handle_t training_handle,
+                                                          bool trainable_only);
+
+/**
+ * Copy all parameters to a contiguous buffer held by the argument parameters_buffer
+ *
+ * User is responsible for allocating and freeing resources used by the parameters_buffer.
+ * Parameter ordering is preserved.
+ *
+ * @param training_handle handle of the training session
+ * @param parameters_buffer pre-allocated OrtValue buffer to copy onto. Must be same size as results of
+ *                          GetParametersSize api call
+ * @param trainable_only whether to skip non-trainable parameters
+ */
+ort_tensor_handle_t EMSCRIPTEN_KEEPALIVE OrtTrainingCopyParametersToBuffer(ort_training_session_handle_t training_handle,
+                                                                           ort_tensor_handle_t parameters_buffer,
```

Comment:
it's a better way to add a parameter for the size of the buffer in the API, and implementation checks on the size. when passing raw buffer always pass pointer + size.

Reply:
The C API method for this function doesn't take in a size_t, so just the buffer will be passed in